### PR TITLE
Change how overlapped works for accepts

### PR DESCRIPTION
**Pull request by davidfowl**
- Make the AsyncAcceptContext an overlapped instead of using Preallocated overlapped. This removes all of the ref counting as there's no shared ownership in the accept loops.
- This also simplifies the creation and disposal logic as there's a single overlapped used for all accepts per context.

### CODE REVIEW

**Comment by davidfowl** [comment]
@kouvel @jkotas @MichalStrehovsky It occurs to me that it the overlapped implementation in the NativeAOT branch doesn't quite work like the CoreCLR implementation and relies on the windows thread pool to dequeue IO. In particular, nobody is calling `StartThreadpoolIo`...

**Comment by jkotas** [comment]
Yes, the NativeAOT branch is using Windows built-in threadpool on Windows. Do you see a problem with it?

Using built-in threadpool has pros and cons. The upside is less code, sharing threadpool with unmanaged components in the same process, and better (at least in theory) self-tunning since the OS threadpool uses more OS internal data-points. https://channel9.msdn.com/Shows/Going+Deep/Inside-Windows-8-Pedro-Teixeira-Thread-pool explains it in more detail. The downside is that we do not have an easy way to tweak the behavior for .NET when we find that OS heuristics do not work well.

**Comment by davidfowl** [comment]
I think we should use the same thread pool by default an opt into OS intrinsics when we have to. However, overlapped IO is a windows specific API so that argument doesn't hold as strongly here. I think the APIs that exist today around overlapped IO would need to be deprecated if we can't make them work with the windows thread pool APIs. My biggest concern is fixing a bug and dropping potential support for  NativeAOT because I changed the APIs that I'm using.

I filed an issue here https://github.com/dotnet/runtimelab/issues/491 to discuss further.

**Review by Tratcher — APPROVED** [review]
Ambitious. Are there other examples where something has derived from Overlapped?

**Comment by halter73** [comment]
How does this fail in NativeAOT scenarios? Does it throw, leak or hang? Is there a good way to test it?

If ThreadPoolBoundHandle._threadPoolHandle were exposed similarly to ThreadPoolBoundHandle.Handle, could we fix the issue by p/invoking StartThreadpoolIo/CancelThreadpoolIo ourselves in NativeAOT scenarios? This is low-level enough maybe exposing the handle is the best way to get the flexibility we want.

**Comment by jkotas** [comment]
> How does this fail in NativeAOT scenarios? Does it throw, leak or hang?

It will probably result into a fatal crash or hang.

We have more discussion about this in https://github.com/dotnet/runtimelab/issues/491#issuecomment-749375356.

TLDR: `ThreadPoolBoundHandle` was explicitly designed to be independent on the legacy `Overlapped` type. The problem is that this change assumes that `ThreadPoolBoundHandle` uses the legacy `Overlapped` type as its internal implementation. It still happens to be the case in the current CoreCLR implementation, but it is not the case in the native aot implementation. There is no reliable way to check for this. The right way to make the improvement in this PR happen is to introduce new APIs on `ThreadPoolBoundHandle` that are not coupled to the legacy `Overlapped` type.

**Comment by davidfowl** [comment]
There are 2 options here:
1. [ ] - Revert this change
2. [ ] - Use ThreadPool.BindHandle directly instead of https://github.com/dotnet/aspnetcore/blob/519d96b6587cc9b20f382d0d6164fdaa0fcffa44/src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs#L104

If I revert this change I'll also need to revert https://github.com/dotnet/aspnetcore/pull/28559 since it makes similar assumptions.
